### PR TITLE
Adds cereal offer endpoint

### PR DESCRIPTION
**Pull request by euthayna**
### Overview
This is a solution for the problem described in this [repository](https://github.com/Superfiliate/hiring/blob/main/the-cereal-offer.md).

Summary: The task is to develop an API endpoint that accepts a JSON input containing cereal items and returns the discounted price for each item, as well as the total price of the cart. This is the input example given in the link above. 
```ruby
{
  "cart": {
    "reference": "2d832fe0-6c96-4515-9be7-4c00983539c1",
    "lineItems": [
      { "name": "Peanut Butter", "price": "39.0", "collection": "BEST-SELLERS" },
      { "name": "Banana Cake", "price": "34.99", "collection": "DEFAULT" },
      { "name": "Cocoa", "price": "34.99", "collection": "KETO" },
      { "name": "Fruity", "price": "32", "collection": "DEFAULT" }
    ]
  }
}
```

### Requirements
These requirements below were collected from the exercise repo description.

- [x] return the _discounted price_ for each individual item  
- [x] return the _final discounted price_ for the entire cart

**Discount Calculation**
 - [x] when buying **one box**, **no discount** should be applied
 - [x] when buying **2 cereal boxes**, you **get a 5% discount** **on those two** boxes.
 - [x] when buying **3 cereal boxes**, you **get a 10% discount**.
 - [x] when buying **4 cereal boxes**, you **get a 20%** discount.
 - [x] when buying **5 or more**, you **get a 25%** discount.
 - [x] when buying cereal boxes from the **KETO** **collection** **should be excluded from the discount** offer.
 - [x] The JSON example provided here should be used as the input schema for the API

I assume these other requirements considering the expected happy path. For the correct calculation, we need to have the **price** and **collection** at least. I added the requirement for the **name** too. Even if we do not use directly, does not make sense to return a discount for an item that does not have a name

Taking into account the requirements mentioned above, we can outline the following requirements:
- _cart_ must be present and be a hash
- _lineItems_ keys must be present and an array of hashes
	- price key: 
		- [x] must be present
		- [x] must be a positive number
	- name key:
		- [x] must be present
		- [x] must be a string
	- collection key:
		- [x] must be present
		- [x] must be a string
*Note: In a real task, I would make fewer assumptions as possible. I usually ask questions to clarify and confirm what I don't know.

### Demo
I used [Postman](https://www.postman.com/) tool as the client. Using the JSON input shown above, have the following JSON output:

```ruby
{
    "cart": {
        "reference": "2d832fe0-6c96-4515-9be7-4c00983539c1",
        "lineItems": [
            {
                "name": "Peanut Butter",
                "price": "39.0",
                "collection": "BEST-SELLERS",
                "discounted_price": 35.1
            },
            {
                "name": "Banana Cake",
                "price": "34.99",
                "collection": "DEFAULT",
                "discounted_price": 31.49
            },
            {
                "name": "Cocoa",
                "price": "34.99",
                "collection": "KETO",
                "discounted_price": 34.99
            },
            {
                "name": "Fruity",
                "price": "32",
                "collection": "DEFAULT",
                "discounted_price": 28.8
            }
        ],
        "totalPrice": 130.38
    }
}
```

### A few considerations:
- We could have created a class to encapsulate the line item and handle the discount logic.
- We'd test more edge cases around box combinations. 
- The controller layer validates the **cart**, **reference**, and **lineItems** keys, returning an error and a 422 status code if invalid. The cart class layer checks the **name**, **price**, and **collection** of line items.
	- Looking back, it would be simpler and easier to manage if all validations were centralized. We could achieve this by passing the entire cart hash to the cart class, not just the line items. This approach also aligns with the class name '_cart_'. Alternatively, all validations could return a status code in the controller layer.
- I added the presence of a _reference_ as a criterion, but even if this were required in a real API input, for the execution of this exercise, we do not use a reference for any logic. 
- I couldn't take any functional criteria for this [sentence](https://github.com/Superfiliate/hiring/blob/main/the-cereal-offer.md). In a real task, I would ask and discuss with someone to understand better if this price interval should be represented in the business logic.
    > A box of cereal usually costs anything between $32 and $39.
- I didn't test any large cart size input

### CODE REVIEW

File: app/models/cart.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        errors.add(:base, "Collection can't be blank")
+      elsif !item[:collection].is_a?(String)
+        errors.add(:base, 'Collection must be a string')
+      end
+    end
+  end
+
+  def calculate_discounts
+    @line_items.dup.map do |item|
+      item[:discounted_price] = if eligible_discount?(item)
```
Comment (euthayna) [inline app/models/cart.rb:65]: in this instance, the _snake_case_ convention has been used for the `discounted_price key`, which is the with Ruby on Rails' (RoR) standard naming convention. However, this key is part of the JSON response, introducing a discrepancy in the response naming conventions since JSON typically uses _camelCase_.

This is an inconsistency that I have only just identified upon reviewing the code here.

Comment (victorfeijo) [reply app/models/cart.rb:65]: That makes sense! Ideally, we would have the JSON response following the same format across all keys, would that be camel case or snake case. The anti-pattern here would be to use two conventions on the same JSON object.

Great attitude from your side to identify your own mistake and propose a change to it 👏🏽

File: spec/rails_helper.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
+# This file is copied to spec/ when you run 'rails generate rspec:install'
```
[the remark this answers is not in the document]
Comment (victorfeijo) [reply spec/rails_helper.rb:1]: Awesome, thank you for the heads up 👍🏽

File: app/controllers/carts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  def create
+    cart = Cart.new(cart_params[:lineItems])
+
+    if cart.valid?
+      format_response(cart)
+    else
+      render json: { errors: cart.errors.full_messages }, status: :unprocessable_entity
+    end
+  rescue ActionController::UnpermittedParameters => e
```
Comment (victorfeijo) [inline app/controllers/carts_controller.rb:15]: `#CON` I believe it would be more beneficial to have this error handling on the `ApplicationController` class. Other controllers could also benefit from it. Does that idea make sense to you?

File: app/controllers/carts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+# This controller deals with the cart API requests.
+class CartsController < ApplicationController
+  before_action :check_cart_params, only: :create
+
+  def create
+    cart = Cart.new(cart_params[:lineItems])
+
+    if cart.valid?
```
Comment (victorfeijo) [inline app/controllers/carts_controller.rb:10]: `#PRO` Handling model errors 👍🏽

File: app/controllers/carts_controller.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    render json: { error: e }, status: :unprocessable_entity
+  end
+
+  private
+
+  def cart_params
+    params.require(:cart).permit(:reference, lineItems: %i[name price collection])
+  end
+
+  def check_cart_params
```
Comment (victorfeijo) [inline app/controllers/carts_controller.rb:25]: `#question` What are your thoughts about checking cart params and formatting the final response on the model level instead of doing here in the controller?

File: app/models/cart.rb [hunk 6]
```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+# This class validates the presence and format of the keys we need to calculate discounts as expected.
+# Calutates the discount price and returns the hash with the discounted_price key for each item element and the key total_price.
+class Cart
+  include ActiveModel::Validations
+  # This hash has functions as key to find the appropriate discount rate.
+  # Depending on the state of the cart (e.g how many items there are), we will pick a different discount.
+  DISCOUNT_PROGRESSION = {
```
Comment (victorfeijo) [inline app/models/cart.rb:9]: `#PRO` Nice FP tricks!! lol

File: app/models/cart.rb [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    {
+      line_items: items,
+      total_price: items.map { |item| item[:discounted_price] }.sum.round(2)
+    }
+  end
+
+  private
+
+  def validate_line_items
```
Comment (victorfeijo) [inline app/models/cart.rb:34]: I think another approach for this would be to create another model class `LineItem` and then add those validations inline for each item. Something like:

```ruby
class LineItem
  include ActiveModel::Validations
  
  validates :name, :price, :collection, presence: true
  validates :price, numericality: { greater_than_or_equal_to: 0.0, allow_nil: false }
```

I wouldn't worry much about requiring those values to be a String, but you could probably develop some custom validation as well. With that said, I believe these validations are great for making a more robust API and ensuring users receive a detailed error message of what is possibly wrong. It would also depend on what the customer (API user) needs, and in this specific case, it feels a bit like we're going too far but much better than not adding validations and error messages at all. Does that make sense?

File: app/models/cart.rb [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+      if item[:collection].blank?
+        errors.add(:base, "Collection can't be blank")
+      elsif !item[:collection].is_a?(String)
+        errors.add(:base, 'Collection must be a string')
+      end
+    end
+  end
+
+  def calculate_discounts
+    @line_items.dup.map do |item|
```
Comment (victorfeijo) [inline app/models/cart.rb:64]: `#PRO` Duplicating the original object, avoiding to mutate its original values 👍🏽

File: app/models/cart.rb [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                                else
+                                  item[:price].to_f
+                                end
+      item
+    end
+  end
+
+  # This method is used to find the discount percentage we must apply, regarding
+  # the discountable_items_count output (quantity of items in the cart).
+  def discount_percentage
```
Comment (victorfeijo) [inline app/models/cart.rb:76]: I like this approach because of the functional programming aspect of it, but probably creating a simple hash and handling the 5+ case would be easier to read and also wouldn't require multiple lambda executions with the `find` method. Does that make sense?

File: app/models/cart.rb [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    DISCOUNT_PROGRESSION.find { |condition, _| condition.call(discountable_items_count) }[1]
+  end
+
+  def discountable_items_count
+    @line_items.count { |item| eligible_discount?(item) }
```
Comment (victorfeijo) [inline app/models/cart.rb:81]: `#PRO` nice usage of count as an iterator here 👍🏽

File: app/models/cart.rb (hunk 9 above)
Comment (victorfeijo) [inline app/models/cart.rb:76]: `#CON` This method is being called multiple times (n * lineItems), and could probably have some time of memoization.

File: app/models/cart.rb [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+      end
+    end
+  end
+
+  def calculate_discounts
+    @line_items.dup.map do |item|
+      item[:discounted_price] = if eligible_discount?(item)
+                                  (item[:price].to_f * (1.0 - discount_percentage)).round(2)
+                                else
+                                  item[:price].to_f
```
Comment (victorfeijo) [inline app/models/cart.rb:68]: `question` Any reason to use float instead of decimal?

File: spec/models/cart_spec.rb [hunk 12]
```diff
@@ -0,0 +1,288 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'Cart' do
```
Comment (victorfeijo) [inline spec/models/cart_spec.rb:5]: `#PRO` Great test coverage 🔝

File: spec/requests/carts_spec.rb [hunk 13]
```diff
@@ -0,0 +1,145 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'Cart', type: :request do
```
Comment (victorfeijo) [inline spec/requests/carts_spec.rb:5]: `#PRO` Controller tests, great stuff 🔝

**Review by victorfeijo — APPROVED** [review]
Awesome stuff from your side @euthayna , I've added a few PROS and CONS throughout the code and would love to hear your thoughts on it :)

File: app/controllers/carts_controller.rb (hunk 3 above)
Comment (euthayna) [reply app/controllers/carts_controller.rb:15]: @victorfeijo Definitely, we should treat this error in `ApplicationController`, good catch 👏

File: app/controllers/carts_controller.rb (hunk 5 above)
Comment (euthayna) [reply app/controllers/carts_controller.rb:25]: @victorfeijo Ideally, the model deals with business logic only, not with formatting view stuff. 

Doing that at a controller level makes more sense since params are injected into the controller. I don't see any reason to inject parameters into the model, so the model formatting them. With that said, if I had more time, I would use a `serializer` class to deal with formatting.

Not 100% sure if I understood your question. Please let me know if I didn't answer your question. ☺️

File: app/models/cart.rb (hunk 7 above)
Comment (euthayna) [reply app/models/cart.rb:34]: @victorfeijo 
Yes, absolutely. 

Once I reviewed this solution before opening a PR, I realized I could and should have used a `LineItem` class to handle line item stuff. So could not agree with you more; it makes total sense. I even wrote on the PR description _"We could have created a class to encapsulate the line item..."_ (A few considerations section). 

About validations: truly, this usually depends on the customer's needs. All you said make total sense 🙂

(I choose to do all these validations because this is an interview process, and I prefer to do it more than less since part of the process is I show you more as I can 😅)

File: app/models/cart.rb (hunk 11 above)
Comment (euthayna) [reply app/models/cart.rb:68]: @victorfeijo, not at all. My mistake (_just a bad habit of a project I  worked_). We should use `decimals`.

I even told someone last week: _"When dealing with money is better not to use float since we can have some problems with precision"_.  ✌️

File: app/models/cart.rb (hunk 9 above)
Comment (euthayna) [reply app/models/cart.rb:76]: @victorfeijo you are right. This needs to be improved.

Comment (euthayna) [reply app/models/cart.rb:76]: @victorfeijo Yes, it makes sense 🚀

File: app/controllers/carts_controller.rb (hunk 5 above)
Comment (victorfeijo) [reply app/controllers/carts_controller.rb:25]: Awesome, that makes total sense @euthayna - probably not handling on the model but could move from the controller to a service/serializer that would handle this formatting/parsing logic 👍🏽
